### PR TITLE
Do not insert fp16 cast if used by select-like ops

### LINKED ISSUE #2352: SegmentCandidateFinder (fusion_segment.cpp) asserts on findSegments()
### 🐛 Describe the bug

Original issue:
```
RuntimeError: h.has_value() INTERNAL ASSERT FAILED at "/opt/pytorch/pytorch/third_party/nvfuser/csrc/fusion_segmenter.cpp":2634, please report a bug to PyTorch. Can not find a scheduler to schedule fusion segment
```

SegmentCandidateFinder applies graph mutation in `finalize()` without query scheduler, which leads to segment groups that can't be scheduled later.

There is a hack in the handling of cast in segmenter. https://github.com/csarofeen/pytorch/blob/4c6465de0afbcaf8b9f251b1c119147207d7c120/third_party/nvfuser/csrc/fusion_segmenter.cpp#L759-L765

This unconditional mutation of segmented graph is the root cause of the segmenter failure with `gather` and `index_select`, which doesn't like arbitrarily inserted cast ops on the first tensor argument.

### Versions

TOT devel currently have this issue.
```
commit 4c6465de0afbcaf8b9f251b1c119147207d7c120 (csarofeen/devel)
Author: Kevin Stephano <kevin.stephano@gmail.com>
Date:   Fri Jan 20 14:32:48 2023 -0800

    Fix gather test to include removal of Fusion object usage. (#2349)

```

### PR DESCRIPTION
**Pull request by naoyam**
Fixes #2352

### CODE REVIEW

File: third_party/nvfuser/csrc/fusion_segmenter.cpp [hunk 1]
```diff
@@ -855,6 +851,8 @@ std::vector<SegmentedEdge*> SegmentedFusion::castInputOutputToLowerPrecision(
     // Gather exprs that should not be altered
     std::unordered_set<Expr*> uses_to_keep;
```
Comment (zasdfgbnm) [inline third_party/nvfuser/csrc/fusion_segmenter.cpp]: Is this variable still used?

Comment (naoyam) [reply third_party/nvfuser/csrc/fusion_segmenter.cpp]: No, will remove. Thanks.

File: third_party/nvfuser/csrc/fusion_segmenter.cpp [hunk 2]
```diff
@@ -892,11 +891,31 @@ std::vector<SegmentedEdge*> SegmentedFusion::castInputOutputToLowerPrecision(
...
-        }
-      }
+      uses_to_modify.erase(
+          std::remove_if(
+              uses_to_modify.begin(),
+              uses_to_modify.end(),
+              [&](Expr* edge_val_use_expr) {
+                return !merged_group_exprs.count(edge_val_use_expr);
+              }),
+          uses_to_modify.end());
```
Comment (zasdfgbnm) [inline third_party/nvfuser/csrc/fusion_segmenter.cpp:889]: Did you mean the following?
```C++
std::remove_if(
    uses_to_modify.begin(),
    uses_to_modify.end(),
    [&](Expr* edge_val_use_expr) {
      return !merged_group_exprs.count(edge_val_use_expr);
    }),
```
I don't understand why you uses both `erase` and `remove_if`.

Comment (naoyam) [reply third_party/nvfuser/csrc/fusion_segmenter.cpp:889]: I thought that's necessary to actually erase. https://en.cppreference.com/w/cpp/algorithm/remove 

Does `remove_if` actually change the `end` iterator? I thought that's done by `erase`.

Comment (naoyam) [reply third_party/nvfuser/csrc/fusion_segmenter.cpp:889]:
```
#include <vector>
#include <iostream>
#include <algorithm>

void test1(std::vector<int> x) {
  std::remove_if(x.begin(), x.end(),
                 [](auto i) {
                   return i % 2 == 0;
                 });

  for (auto i: x) {
    std::cout << "test1: " << i << std::endl;
  }
}

void test2(std::vector<int> x) {
  x.erase(std::remove_if(x.begin(), x.end(),
                         [](auto i) {
                           return i % 2 == 0;
                         }), x.end());

  for (auto i: x) {
    std::cout << "test2: " << i << std::endl;
  }
}

int main() {
  std::vector<int> x({1, 2, 3, 4});
  test1(x);
  test2(x);
  return 0;
}
```
Output:
```
test1: 1
test1: 3
test1: 3
test1: 4
test2: 1
test2: 3
```

So, looks like without `erase` the end iterator is not moved.

Comment (zasdfgbnm) [reply third_party/nvfuser/csrc/fusion_segmenter.cpp:889]: Oh, I didn't know that. Thanks for explaining!

**Review by zasdfgbnm — APPROVED** [review]

File: third_party/nvfuser/csrc/fusion_segmenter.cpp (hunk 2 above)
Comment (naoyam) [reply third_party/nvfuser/csrc/fusion_segmenter.cpp:889]: NP. There's `std::erase` in C++20, which would be much simpler.

**Pull request merged**